controllers/topology/machinedeployment_controller.go

```diff
@@ -0,0 +1,161 @@
+// Note: To achieve this the cluster topology controller sets a finalizer to hook into the MachineDeployment deletions.
+type MachineDeploymentReconciler struct {
+	Client           client.Client
+	WatchFilterValue string
+}
+
+func (r *MachineDeploymentReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineDeployment{}).
+		Named("machinedeployment/topology").
+		WithOptions(options).
+		WithEventFilter(predicates.All(ctrl.LoggerFrom(ctx),
+			predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue),
+			predicates.ResourceIsTopologyOwned(ctrl.LoggerFrom(ctx)),
+		)).
+		Complete(r)
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	return nil
+}
+
+// Reconcile deletes referenced templates during deletion of topology-owned MachineDeployments.
+// The templates are only deleted, if they are not used in other MachineDeployments or MachineSets which are not in deleting,
+// i.e. the templates would otherwise be orphaned after the MachineDeployment deletion completes.
+// Additional context:
+// * MachineDeployment deletion:
+//   * MachineDeployments are deleted and garbage collected first (without waiting until all MachineSets are also deleted).
+//   * After that, deletion of MachineSets is automatically triggered by Kubernetes based on owner references.
+// Note: We assume templates are not reused by different MachineDeployments, which is only true for topology-owned
+//       MachineDeployments.
+func (r *MachineDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the MachineDeployment instance.
+	md := &clusterv1.MachineDeployment{}
+	if err := r.Client.Get(ctx, req.NamespacedName, md); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get MachineDeployment/%s", req.NamespacedName.Name)
+	}
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, md.Namespace, md.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Return early if the object or Cluster is paused.
+	if annotations.IsPaused(cluster, md) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// We don't have to set the finalizer, as it's already set during MachineDeployment creation
+	// in the cluster topology controller.
+
+	// Handle deletion reconciliation loop.
+	if !md.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, md)
+	}
+
+	// Nothing to do.
+	return ctrl.Result{}, nil
+}
+
+func (r MachineDeploymentReconciler) reconcileDelete(ctx context.Context, md *clusterv1.MachineDeployment) (ctrl.Result, error) {
```

Comment:
Everything below here is unique to this reconciler, everything above just basic setup